libsync/SyncMaster.cpp

```diff
@@ -499,20 +506,29 @@ void SyncMaster::maintainPeersConnection()
     bool hasMyself = false;
     for (auto const& member : minerOrObserver)
     {
+        /// find active peers
         if (activePeers.find(member) != activePeers.end() && member != m_nodeId)
+        {
             memberSet.insert(member);
-
+        }
         hasMyself |= (member == m_nodeId);
     }
 
-    // Delete uncorrelated peers
-    NodeIDs nodeIds = m_syncStatus->peers();
-    for (NodeID const& id : nodeIds)
+    // Delete uncorrelated peers(only if the node the miner or observer, check the identities of
+    // other peers)
+    if (hasMyself)
     {
-        if (memberSet.find(id) == memberSet.end())
-            m_syncStatus->deletePeer(id);
+        NodeIDs nodeIds = m_syncStatus->peers();
+        for (NodeID const& id : nodeIds)
+        {
+            if (memberSet.find(id) == memberSet.end())
+            {
+                m_syncStatus->deletePeer(id);
```

Comment:
break here ? to avoid traversing the list

Reply:
Add `break` here will change the logic.